src/main/lib/icon/storage/index.js

```diff
@@ -0,0 +1,9 @@
+import IconFileStorage from './IconFileStorage';
+import electron from 'electron';
+import path from 'path';
+
+const app = electron.app || electron.remote.app;
+const iconPath = path.join(app.getPath('userData'), 'icons');
+
+const storage = new IconFileStorage(iconPath);
+export default storage;
```

Comment:
Btw, here a practical example of that `index` pattern we discussed.
If you still don't like it like this, I'm open to discuss other approaches.
IMO it's just a great way of hiding the wirings inside a folder + testing very easily the `IconFileStorage`

Reply:
I'm still not sold on it.. I don't see immediate benefit from using it and I feel it overcomplicates the file structure of the application. In your example you return some-what of a singleton as you're instantiating the class and providing an arg to it.. but this could all be done in-module imo. For instance:

```javascript
let __sharedInstance;

class MyClass {}

export default MyClass;

export function getSharedInstance() {
  if (!__sharedInstance) {
    __sharedInstance = new MyClass(stuff);
  }
  return __sharedInstance;
}
```

The class itself is easily testable and mockable at this point. The only reason that you'd want to intercept `__sharedInstance` here is for integration testing. For that I'd argue that when integration testing, you should mock as little as possible to test the actual platform.. especially if you're going to the trouble of integration testing. Also, if you are going to have to change it to point elsewhere for some reason, you could inject said dependency upstream or provide a setter for the singleton. This latter way is obviously more ugly but integration testing sometimes needs a more complex harness for it to work.

Tldr: I don't think that, for the sake of integration testing, that it's necessary to add index files for directories 😅

Also, I may have completely misunderstood your point.. if so I apologise 😬

Reply:
Im good with both approaches

Reply:
@perry-mitchell I see your point in that actually thanks to JS modules supporting default + secondary exports, it is possible to use this in different ways to define in one single JS file both the "pure" module (unit testable) + the instantiated one (dependent on specific config or on other modules which you would want to replace in your test).

I guess in the end it comes to a matter of taste of how to apply the same idea. I like better somehow to separate very clearly in different files the pure module from it's specific instantiation/configuration.

In this case anyway, the point of this was to expose the pure non-configured `IconFileStorage` in one place (so in the test I pass my own specific path) and in another place provide both its real config (the Electron path) and make it a singleton.
It is totally about unit testing it (considering it unit even if using the filesystem).

Because here it is really not going to be used anywhere out of Iconographer, I could actually remove the `index` and do that part inside the `iconographer.js` itself. What about that?
But otherwise, an alternative for this could be what you suggest.

Reply:
Pd. I actually love how in JS you have such freedom of going with so many different approaches in this aspect 👍

Reply:
Pd2. I love discussing these meta aspects of code... maybe you noticed 😀

Reply:
> Because here it is really not going to be used anywhere out of Iconographer, I could actually remove the index and do that part inside the iconographer.js itself. What about that?

> But otherwise, an alternative for this could be what you suggest.

Isn't that the same as what I suggested? 😅

I was suggesting having the singleton/instance getter inside the relevant module file.

> I actually love how in JS you have such freedom of going with so many different approaches in this aspect

Me too! It's sometimes a burden as it means there's more ways in which you can hurt yourself when it comes to maintainability, but usually it's a good thing.

> I love discussing these meta aspects of code... maybe you noticed

I did! Haha.. and so do we. Always great to have fresh eyes on what we're building too!